src/EfuncMsg.chpl

```diff
@@ -1375,4 +1500,58 @@ module EfuncMsg
       }
       return C;
     }
+//}
+//}
+    proc putmask_helper (mask : [?D1] bool, ref A : [D1] ?ta, Values : [?D2] ?tv) {
+
+//  Note, added casting to the output, to accomodate the various combinations of
+//  types that are allowed by np.putmask.  Invalid pairings of types ta, tv must be 
+//  caught and flagged by the calling function.  This proc does no type checking.
+
+      if A.size == Values.size { // then their distributions match
+
+	forall element in A.domain {
+		if mask[element] then A[element] = Values[element]:ta ;
+	}	
+
+      } else if A.size < Values.size { // then prune Values so the distributions match
+
+	var Pruned_Values = Values[0..A.size-1] ;
+	forall element in A.domain {
+		if mask[element] then A[element] = Pruned_Values[element]:ta ;
+	}	
+
+//  For the remaining cases, we know Values.size < A.size
+
+      } else {
+	var A_subset_size = (A.size/numLocales):int ;
+	if A.size%numLocales == 0 then A_subset_size += 1 ;
+
+//  If Values.size < A_subset_size, then we may as well make a copy of Values on each locale
+
+	if Values.size < A_subset_size {
+		coforall loc in Locales do on loc {
+			var local_Values = Values ;
+			forall element in A.localSubdomain() {
+				if mask[element] then A[element] = (local_Values[element%local_Values.size]):ta ;
+			}
+		}
```

Comment:
I pulled down this PR to check the comm numbers with multi-locale for the test cases u wrote. They all seemed pretty reasonable except for this one

I was seeing these numbers with 3 locales and problem size `10**6`:
```
time = 2.31879
| locale |   get | put | execute_on | execute_on_nb |
| -----: | ----: | --: | ---------: | ------------: |
|      0 | 89048 |   4 |          0 |            14 |
|      1 | 88547 |   2 |          1 |            12 |
|      2 | 89146 |   2 |          1 |            12 |
```

it's because `var local_Values = Values ;` isn't actually making a local copy, it's making a copy that's a dist array

update that line to specify the domain:
```chapel
var local_Values : [0..Values.size-1] tv = Values ;
```
and with the same problem size and number of locales, we get the comm numbers we expect:
```
time = 0.0183921
communication:
| locale | get | put | execute_on_nb |
| -----: | --: | --: | ------------: |
|      0 |   6 |   2 |             4 |
|      1 |   6 |   2 |             2 |
|      2 |   6 |   2 |             2 |
```

Reply:
I made the change, and would appreciate some team discussion about this at our next opportunity.